File: airesources/JavaScript/hlt/Planet.js
const Entity = require('./Entity');

class Planet extends Entity {
    /**
     * @param {GameMap} gameMap map this planet belongs to
     * @param {object} params planet information
     */
    constructor(gameMap, params) {
        super(params);
        this._params = params;
        this._gameMap = gameMap;
    }

    /**
     * Determines if the planet has an owner.
     * @returns {boolean} true if planet is owned
     */
    isOwned() {
        return this.ownerId !== null &&
            typeof this.ownerId !== 'undefined';
    }

    /**
     * Determines if the planet owner is you.
     * @returns {boolean} true if planet is owned by you
     */
    isOwnedByMe() {
        return this.ownerId === this._gameMap.myPlayerId;
    }

    /**
     * Determines if the planet owner is not you.
     * @returns {boolean} true if planet is owned by you
     */
    isOwnedByEnemy() {
        return this.isOwned() && this.ownerId !== this._gameMap.myPlayerId;
    }

    /**
     * Determines if the planet is free.
     * @returns {boolean} true if planet is free
     */
    isFree() {
        return this.ownerId === null ||
            typeof this.ownerId === 'undefined';
    }

    get ownerId() {
        return this._params.ownerId;
    }

    /**
     * number of docking spots in this planet
     * @return {number} number of docking spots in this planet
     */
    get dockingSpots() {
        return this._params.dockingSpots;
    }

    /**
     * determines if there is docking spot available
     * @return {boolean} true if there is a docking spot
     */
    hasDockingSpot() {
        return this.numberOfDockedShips < this._params.dockingSpots;
    }

    get currentProduction() {
        return this._params.currentProduction;
    }

    get remainingProduction() {
        return this._params.remainingProduction;
    }

    /**
     * docked ship ids
     * @returns {number[]}
     */
    get dockedShipIds() {
        return this._params.dockedShipIds;
    }

    /**
     * number of docked ships
     * @returns {number}
     */
    get numberOfDockedShips() {
        return this.dockedShipIds.length;
    }

    /**
     * docked ships instances
     * @returns {Ship[]}
     */
    get dockedShips() {
        return this._gameMap.shipsByIds(this.dockedShipIds);
    }

    toString() {
        return 'planer. ' + (this.ownerId ? 'owner id: ' + this.ownerId : 'no owner') + ': ' + JSON.stringify(this._params);
    }
}

module.exports = Planet;

File: airesources/Swift/Sources/hlt/Planet.swift
import Foundation

public struct Planet: Entity, HLTDeserializable {
    /// The id of this planet
    public let id: Int
    
    /// The x coordinate of the planet
    public let x: Double
    
    /// The y coordinate of the planet
    public let y: Double
    
    /// The health of the planet
    public let health: Int
    
    /// The radius of the planet
    public let radius: Double
    
    /// The number of docking spots that the planet has
    public let dockingSpots: Int
    
    /// The current production of the planet
    public let currentProduction: Int
    
    /// The remaining production of the planet
    public let remainingProduction: Int
    
    /// Whether the planet is owned or not
    public let owned: Bool
    
    /// If the planet is owned, the id of the player that owns it, otherwise, nil
    public let ownerId: Int?
    
    /// The number of ships docked at the planet
    public let dockedShipCount: Int
    
    /// The ids of the ships docked at the planet
    public let dockedShipIds: [Int]
    
    static func deserialize(_ tokens: TokenStack) -> Planet {
        let id = Int(tokens.pop())!
        let x = Double(tokens.pop())!
        let y = Double(tokens.pop())!
        let health = Int(tokens.pop())!
        let radius = Double(tokens.pop())!
        let dockingSpots = Int(tokens.pop())!
        let currentProduction = Int(tokens.pop())!
        let remainingProduction = Int(tokens.pop())!
        let owned = (Int(tokens.pop())! != 0) ? true: false
        let ownerId = Int(tokens.pop())!
        let dockedShipCount = Int(tokens.pop())!
        var dockedShipIds = [Int]()
        
        for _ in 0..<dockedShipCount {
            dockedShipIds.append(Int(tokens.pop())!)
        }
        
        return Planet(id: id,
                      x: x,
                      y: y,
                      health: health,
                      radius: radius,
                      dockingSpots: dockingSpots,
                      currentProduction: currentProduction,
                      remainingProduction: remainingProduction,
                      owned: owned,
                      ownerId: owned ? ownerId : nil,
                      dockedShipCount: dockedShipCount,
                      dockedShipIds: dockedShipIds)
    }
}
